### PR TITLE
Adds subscription fulfillment module and course individual payments implementation

### PR DESCRIPTION
**Pull request by rehan99000**
This PR adds the following:
- Fulfillment module for publishing user subscription to LMS
- Course individual payments implementation (ecommerce side) (frontend and backend)
- Bug fixes and cleanup

**JIRA:**
https://edlyio.atlassian.net/browse/EDLY-1818

### CODE REVIEW

File: ecommerce/extensions/api/serializers.py [hunk 1]
```diff
@@ -185,7 +185,8 @@ def get_value(self, obj):
             vouchers = obj.value.vouchers.all()
             serializer = VoucherSerializer(vouchers, many=True, context={'request': request})
             return serializer.data
-        return obj.value
+        value = obj.value.option if obj.attribute.type == 'option' else obj.value
```
Comment (rehan99000) [inline ecommerce/extensions/api/serializers.py:188]: required for option attributes of subscription for order history view in LMS

**Review by ziafazal — CHANGES_REQUESTED** [review]
@rehan99000 overall looks good few minor things to look into.

File: ecommerce/subscriptions/api/v2/views.py [hunk 2]
```diff
@@ -56,3 +56,13 @@ def get_serializer_context(self):
         context = super(SubscriptionViewSet, self).get_serializer_context()
         context['partner'] = get_partner_for_site(self.request)
         return context
+
+    @list_route(methods=['get'])
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/views.py:60]: @ziafazal The `POST` call requires a csrf token that for some reason is not sent when I call it from the backbone. Should I use `POST` with csrf exempt?

File: ecommerce/static/js/views/confirmation_dialog_view.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+				context: this,
+				type: 'GET',
+				url: window.location.origin +
+					'/api/v2/subscriptions/toggle_course_individual_payments/',
+				success: this.onSuccess,
+			});
+		},
+
+		onSuccess: function (data) {
+            var course_individual_payments_button = $('[name=course-individual-payments]');
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/static/js/views/confirmation_dialog_view.js]: It was showing fine with me. still auto formatted

File: ecommerce/static/js/views/confirmation_dialog_view.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		},
+
+		confirmToggle: function (event) {
+			event.preventDefault();
+			this.toggleIndividualCoursePaymentsData();
+		},
+
+		toggleIndividualCoursePaymentsData: function () {
+			Backbone.ajax({
+				context: this,
```
Comment (ziafazal) [inline ecommerce/static/js/views/confirmation_dialog_view.js:28]: @rehan99000 
You pass CSRF token like this
```
  headers: {
                        'X-CSRFToken': Cookies.get('ecommerce_csrftoken')
                    },
```

Comment (rehan99000) [reply ecommerce/static/js/views/confirmation_dialog_view.js:28]: done

File: ecommerce/subscriptions/api/v2/serializers.py [hunk 5]
```diff
@@ -77,11 +78,18 @@ def get_display_order(self, product):
         """
         return product.attr.subscription_display_order
 
+    def get_course_individual_payments(self, product):
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/api/v2/serializers.py:81]: `get_course_payments_enabled`

File: ecommerce/subscriptions/api/v2/serializers.py [hunk 6]
```diff
@@ -77,11 +78,18 @@ def get_display_order(self, product):
...
+        """
+        site_configuration = self.context['request'].site.siteconfiguration
+        return site_configuration.enable_course_individual_payments
+
     class Meta:
         model = Product
         fields = [
             'id', 'title', 'date_created', 'subscription_type', 'subscription_actual_price', 'subscription_price',
-            'subscription_status', 'display_order', 'partner_sku',
+            'subscription_status', 'display_order', 'partner_sku', 'course_individual_payments'
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/api/v2/serializers.py:92]: `course_payments_enabled`

File: ecommerce/subscriptions/modules.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                )
+                order.notes.create(message='Fulfillment of order failed due to a network problem.', note_type='Error')
+                line.set_status(LINE.FULFILLMENT_NETWORK_ERROR)
+            except Timeout:
+                logger.error(
+                    "Unable to fulfill line [%d] of order [%s] due to a request time out", line.id, order.number
+                )
+                order.notes.create(message='Fulfillment of order failed due to a request time out.', note_type='Error')
+                line.set_status(LINE.FULFILLMENT_TIMEOUT_ERROR)
+        logger.info("Finished fulfilling 'Subscription' product types for order [%s]", order.number)
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/modules.py:142]: nit: add a new line before de-indent

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added my feedback.

File: ecommerce/settings/_oscar.py [hunk 8]
```diff
@@ -81,6 +81,7 @@
     'ecommerce.extensions.fulfillment.modules.EnrollmentCodeFulfillmentModule',
     'ecommerce.extensions.fulfillment.modules.CourseEntitlementFulfillmentModule',
     'ecommerce.extensions.fulfillment.modules.DonationsFromCheckoutTestFulfillmentModule',
+    'ecommerce.subscriptions.modules.SubscriptionFulfillmentModule',
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/settings/_oscar.py:85]: done

File: ecommerce/core/models.py [hunk 9]
```diff
@@ -195,6 +195,12 @@ class SiteConfiguration(models.Model):
         default=False
     )
 
+    enable_course_individual_payments = models.BooleanField(
+        verbose_name=_('Enable Course Individual Payments'),
+        help_text=_('Flag to check if site wise course individual payments are enabled'),
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/core/models.py]: done

File: ecommerce/settings/_oscar.py [hunk 10]
```diff
@@ -275,6 +276,7 @@
 
 # Default timeout for Enrollment API calls
 ENROLLMENT_FULFILLMENT_TIMEOUT = 7
+USER_SUBSCRIPTION_FULFILLMENT_TIMEOUT = 7
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/settings/_oscar.py:281]: done

File: ecommerce/core/models.py [hunk 11]
```diff
@@ -195,6 +195,12 @@ class SiteConfiguration(models.Model):
         default=False
     )
 
+    enable_course_individual_payments = models.BooleanField(
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/core/models.py]: done

File: ecommerce/static/js/views/subscription_list_view.js [hunk 12]
```diff
@@ -105,8 +110,18 @@ define([
                 }
             },
 
+            toggleIndividualCoursePaymentsButtonText: function() {
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/static/js/views/subscription_list_view.js]: done

File: ecommerce/static/templates/individual_payments_action_modal.html [hunk 13]
```diff
@@ -0,0 +1,20 @@
+<div class="modal fade" id="individualPaymentsActionModal" tabindex="-1" role="dialog"
```
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/static/templates/individual_payments_action_modal.html]: done

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 5 above)
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/serializers.py]: done

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 6 above)
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/serializers.py]: done

File: ecommerce/subscriptions/modules.py (hunk 7 above)
Comment (rehan99000) [reply ecommerce/subscriptions/modules.py:143]: done

File: ecommerce/extensions/api/serializers.py (hunk 1 above)
Comment (rehan99000) [reply ecommerce/extensions/api/serializers.py]: done

File: ecommerce/static/js/views/confirmation_dialog_view.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rehan99000) [reply ecommerce/static/js/views/confirmation_dialog_view.js]: done

File: ecommerce/core/migrations/0050_add_course_individual_flag_in_siteconfiguration.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Comment (zubair-arbi) [inline ecommerce/core/migrations/0050_add_course_individual_flag_in_siteconfiguration.py:1]: Please delete this migration and create again as the name of this migration doesn't seem right `0050_add_course_individual_flag_in_siteconfiguration.py`

File: ecommerce/subscriptions/api/v2/serializers.py [hunk 15]
```diff
@@ -77,11 +78,18 @@ def get_display_order(self, product):
         """
         return product.attr.subscription_display_order
 
+    def get_course_payments(self, product):
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/api/v2/serializers.py:81]: rename it to `is_course_payments_enabled`

File: ecommerce/subscriptions/modules.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+                    line.set_status(LINE.FULFILLMENT_SERVER_ERROR)
+            except ConnectionError:
+                logger.error(
+                    "Unable to fulfill line [%d] of order [%s] due to a network problem", line.id, order.number
+                )
+                order.notes.create(message='Fulfillment of order failed due to a network problem.', note_type='Error')
+                line.set_status(LINE.FULFILLMENT_NETWORK_ERROR)
+            except Timeout:
+                logger.error(
+                    "Unable to fulfill line [%d] of order [%s] due to a request time out", line.id, order.number
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/modules.py:138]: Be consistent with using single quotes.

File: ecommerce/static/js/views/confirmation_dialog_view.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+				},
+				url:
+					window.location.origin +
+					'/api/v2/subscriptions/toggle_course_payments/',
+				success: this.onSuccess,
+			});
+		},
+
+		onSuccess: function (data) {
+			var course_payments_button = $('[name=course-payments]');
```
Comment (zubair-arbi) [inline ecommerce/static/js/views/confirmation_dialog_view.js:42]: Javascript recommends camel case for declaring variable names.
Rename it to `coursePaymentsButton`

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added some more feedback.

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 15 above)
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/serializers.py:81]: This is a serializer method field get method, it needs to have the `get_` prefix.

File: ecommerce/subscriptions/modules.py (hunk 16 above)
Comment (rehan99000) [reply ecommerce/subscriptions/modules.py:138]: done

File: ecommerce/static/js/views/confirmation_dialog_view.js (hunk 17 above)
Comment (rehan99000) [reply ecommerce/static/js/views/confirmation_dialog_view.js:42]: done

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 15 above)
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/serializers.py:81]: to change this, I would have to mention it explicitly in the field declaration. Do u want me to do that?

File: ecommerce/core/migrations/0050_add_course_individual_flag_in_siteconfiguration.py (hunk 14 above)
Comment (rehan99000) [reply ecommerce/core/migrations/0050_add_course_individual_flag_in_siteconfiguration.py:1]: done

File: ecommerce/static/templates/subscription_list.html [hunk 18]
```diff
@@ -2,7 +2,8 @@
     <h1 class="hd-1 emphasized">
         <%- gettext('Subscriptions') %>
         <div class="pull-right">
-            <a href="/subscriptions/new/" class="btn btn-primary btn-small"><%- gettext('Disable Individual Course Payments') %></a>
+            <a href = '#' class="btn btn-primary btn-small" name="course-payments" data-toggle="modal"
```
Comment (ziafazal) [inline ecommerce/static/templates/subscription_list.html:5]: @rehan99000 this href tag has formatting issues like spaces around equal operation and why we don't have display text in it?.

File: ecommerce/subscriptions/modules.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+                )
+                order.notes.create(message='Fulfillment of order failed due to a request time out.', note_type='Error')
+                line.set_status(LINE.FULFILLMENT_TIMEOUT_ERROR)
+
+        logger.info('Finished fulfilling 'Subscription' product types for order [%s]', order.number)
+        return order, lines
+
+    def revoke_line(self, line):
+        try:
+            logger.info('Attempting to revoke fulfillment of Line [%d]...', line.id)
```
Comment (ziafazal) [inline ecommerce/subscriptions/modules.py:148]: which code revokes fulfillment of line?

**Review by ziafazal — APPROVED** [review]
@rehan99000 Overall looks good. One question and one suggestion.

File: ecommerce/subscriptions/modules.py (hunk 19 above)
Comment (rehan99000) [reply ecommerce/subscriptions/modules.py:148]: @ziafazal https://github.com/edly-io/ecommerce/blob/a2ee36711311ad65a7a966dd3188296b6b8bf4ce/ecommerce/extensions/fulfillment/api.py#L57

File: ecommerce/static/templates/subscription_list.html (hunk 18 above)
Comment (rehan99000) [reply ecommerce/static/templates/subscription_list.html:5]: display text on this is set through jquery as we need to display `Enable course payments`/`Disable course payments` according to the flag value in site configuration. format issue fixed.

**Review by asadmanzoor93 — COMMENTED** [review]
@rehan99000 is there a separate PR for unit tests covering this functionality?

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 15 above)
Comment (zubair-arbi) [reply ecommerce/subscriptions/api/v2/serializers.py:81]: Lets change it to `get_is_course_payments_enabled` and in field declaration use `is_course_payments_enabled`

File: ecommerce/extensions/api/serializers.py [hunk 20]
```diff
@@ -185,7 +185,11 @@ def get_value(self, obj):
             vouchers = obj.value.vouchers.all()
             serializer = VoucherSerializer(vouchers, many=True, context={'request': request})
             return serializer.data
-        return obj.value
+        if obj.attribute.type == 'option':
+            value = obj.value.option
+        else:
+            value = obj.value
+        return value
```
Comment (zubair-arbi) [inline ecommerce/extensions/api/serializers.py:192]: nit: add a new line before code de-indent

File: ecommerce/subscriptions/modules.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+    def get_supported_lines(self, lines):
+        """
+        Return a list of lines that can be fulfilled through subscription.
+
+        Checks each line to determine if it is a "Subscription". Subscription are fulfilled by creating a
+        user subscription record in the LMS, which is the sole functionality of this module. Any Subscription product
+        will be returned as a supported line.
+
+        Args:
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/modules.py:54]: Always use full word `Arguments` (https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#docstrings)

File: ecommerce/subscriptions/modules.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    def fulfill_product(self, order, lines, email_opt_in=False):
+        """
+        Fulfills the purchase of a "Subscription" by creating a user subscription record in the LMS.
+
+        Uses the order and the lines to create relevant user subscription records on the LMS side.
+        May result in an error if the User Subscription API cannot be reached, or if there is
+        additional business logic errors when trying to create the user subscription record.
+
+        Args:
+            order (Order): The Order associated with the lines to be fulfilled. The user associated with the order
```
Comment (zubair-arbi) [inline ecommerce/subscriptions/modules.py:71]:
```
The Order associated with the lines to be fulfilled. The user associated with the order
is presumed to be the student buying the subscription.
```

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
Some small nits. Otherwise changes looks good.
Please add some tests for the subscriptions module

File: ecommerce/subscriptions/modules.py (hunk 22 above)
Comment (rehan99000) [reply ecommerce/subscriptions/modules.py:71]: done

File: ecommerce/subscriptions/modules.py (hunk 21 above)
Comment (rehan99000) [reply ecommerce/subscriptions/modules.py:54]: done. it got left while copying from the other module

File: ecommerce/extensions/api/serializers.py (hunk 20 above)
Comment (rehan99000) [reply ecommerce/extensions/api/serializers.py:192]: done

File: ecommerce/subscriptions/api/v2/serializers.py (hunk 15 above)
Comment (rehan99000) [reply ecommerce/subscriptions/api/v2/serializers.py:81]: done

**Review by zubair-arbi — APPROVED** [review]
Change looks good now 👍 
Good job.

**Comment by rehan99000** [comment]
> @rehan99000 is there a separate PR for unit tests covering this functionality?

Yes there will be a separate PR for all the tests

**Pull request merged**